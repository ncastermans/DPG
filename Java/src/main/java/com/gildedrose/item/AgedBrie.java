package com.gildedrose.item;

public class AgedBrie extends BaseItem {
    public AgedBrie(Item item) {
        super(item);
    }

    @Override
    protected void updateQuality() {
        int qualityDelta = (item.sellIn > 0) ? QUALITY_DELTA : QUALITY_DELTA * 2;
        item.quality = item.quality + qualityDelta;
    }
}
