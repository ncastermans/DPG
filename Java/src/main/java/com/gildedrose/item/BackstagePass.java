package com.gildedrose.item;

public class BackstagePass extends BaseItem {
    public BackstagePass(Item item) {
        super(item);
    }

    @Override
    protected void updateQuality() {
        if (item.sellIn < 0) {
            item.quality = 0;
            return;
        }

        int qualityDelta = 1;

        if (item.sellIn <= 5) {
            qualityDelta = 3;
        } else if (item.sellIn <= 10) {
            qualityDelta = 2;
        }

        item.quality = item.quality + qualityDelta;
    }
}
