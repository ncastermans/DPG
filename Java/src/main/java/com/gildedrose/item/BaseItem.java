package com.gildedrose.item;

public class BaseItem {
    protected final int MAX_QUALITY = 50;
    protected final int MIN_QUALITY = 0;
    protected final int SELL_IN_DELTA = 1;
    protected final int QUALITY_DELTA = 1;

    protected Item item;

    public BaseItem(Item item) {
        this.item = item;
    }

    public void update() {
        updateSellIn();
        updateQuality();
        validateQuality();
    }

    protected void updateQuality() {
        int qualityDelta = item.sellIn >= 0 ? QUALITY_DELTA : QUALITY_DELTA * 2;
        item.quality = Math.min(Math.max(item.quality - qualityDelta, MIN_QUALITY), MAX_QUALITY);
    }

    protected void updateSellIn() {
        item.sellIn -= SELL_IN_DELTA;
    }

    protected void validateQuality() {
        item.quality = Math.min(Math.max(item.quality, MIN_QUALITY), MAX_QUALITY);
    }
}
