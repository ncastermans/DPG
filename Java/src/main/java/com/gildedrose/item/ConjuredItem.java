package com.gildedrose.item;

public class ConjuredItem extends BaseItem {
    private final int QUALITY_DELTA = 2;

    public ConjuredItem(Item item) {
        super(item);
    }

    @Override
    protected void updateQuality() {
        int qualityDelta = item.sellIn > 0 ? QUALITY_DELTA : QUALITY_DELTA * 2;
        item.quality = item.quality - qualityDelta;
    }
}
