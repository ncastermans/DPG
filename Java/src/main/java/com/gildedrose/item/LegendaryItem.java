package com.gildedrose.item;

public class LegendaryItem extends BaseItem {

    protected int MAX_QUALITY = 80;

    public LegendaryItem(Item item) {
        super(item);
    }

    @Override
    protected void updateQuality() {
        item.quality = MAX_QUALITY;
    }

    @Override
    protected void updateSellIn() {
        // No need to update sell-in for legendary items
    }

    @Override
    protected void validateQuality() {
        item.quality = MAX_QUALITY;
    }
}
