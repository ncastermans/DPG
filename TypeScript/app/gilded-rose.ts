import { Item } from './models/item';
import { ItemVisitor } from './visitors/item-visitor';
import { ItemVisitorImpl } from './visitors/item-visitor-impl';

export class GildedRose {
  items: Item[];
  private visitor: ItemVisitor;

  constructor(items: Item[]) {
    this.items = items;
    this.visitor = new ItemVisitorImpl();
  }

  updateQuality(): Item[] {
    for (const item of this.items) {
      switch (item.name) {
        case 'Aged Brie':
          this.visitor.visitAgedBrie(item);
          break;
        case 'Backstage passes to a TAFKAL80ETC concert':
          this.visitor.visitBackstagePass(item);
          break;
        case 'Sulfuras, Hand of Ragnaros':
          this.visitor.visitLegendaryItem(item);
          break;
        case 'Conjured Mana Cake':
          this.visitor.visitConjuredItem(item);
          break;
        default:
          this.visitor.visitBaseItem(item);
      }
    }

    return this.items;
  }
}
