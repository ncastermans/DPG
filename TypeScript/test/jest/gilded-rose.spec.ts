import { GildedRose } from '@/gilded-rose';
import { Item } from '@/models/item';

describe('GildedRose', () => {
  it('should match snapshot', () => {
    console.log = jest.fn(); // Mock console.log to prevent output during test

    const items = [
      new Item('+5 Dexterity Vest', 10, 20),
      new Item('Aged Brie', 2, 0),
      new Item('Elixir of the Mongoose', 5, 7),
      new Item('Sulfuras, Hand of Ragnaros', 0, 80),
      new Item('Sulfuras, Hand of Ragnaros', -1, 80),
      new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20),
      new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49),
      new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49),
      new Item('Conjured Mana Cake', 3, 6),
    ];

    const gildedRose = new GildedRose(items);

    let days = 30;

    for (let i = 0; i < days + 1; i++) {
      gildedRose.updateQuality();
    }

    // Compare the console.log output with the snapshot
    expect(console.log).toMatchSnapshot();
  });
});
