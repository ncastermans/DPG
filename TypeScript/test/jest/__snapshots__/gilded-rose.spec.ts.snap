// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GildedRose should match snapshot 1`] = `[MockFunction]`;
